v2/googlecloud-to-googlecloud/src/main/java/com/google/cloud/teleport/v2/transforms/WriteDataChangeRecordsToAvro.java

```diff
@@ -46,91 +46,149 @@ public class WriteDataChangeRecordsToAvro {
+                  record.getMetadata().getPartitionToken(),
+                  timestampToMicros(record.getMetadata().getRecordTimestamp()),
+                  timestampToMicros(record.getMetadata().getPartitionStartTimestamp()),
+                  timestampToMicros(record.getMetadata().getPartitionEndTimestamp()),
+                  timestampToMicros(record.getMetadata().getPartitionCreatedAt()),
+                  record.getMetadata().getPartitionScheduledAt() == null
+                      ? 0
+                      : timestampToMicros(record.getMetadata().getPartitionScheduledAt()),
+                  record.getMetadata().getPartitionRunningAt() == null
+                      ? 0
+                      : timestampToMicros(record.getMetadata().getPartitionRunningAt()),
+                  timestampToMicros(record.getMetadata().getQueryStartedAt()),
+                  timestampToMicros(record.getMetadata().getRecordStreamStartedAt()),
+                  timestampToMicros(record.getMetadata().getRecordStreamEndedAt()),
+                  timestampToMicros(record.getMetadata().getRecordReadAt()),
+                  record.getMetadata().getTotalStreamTimeMillis(),
+                  record.getMetadata().getNumberOfRecordsRead());
+
+      String spannerDatabase = includeSpannerSource() ? spannerDatabase() : null;
+      String spannerInstanceId = includeSpannerSource() ? spannerInstanceId() : null;
+      // Add ChangeStreamMetadata and spanner source info
+      return new com.google.cloud.teleport.v2.DataChangeRecord(
+          partitionToken,
+          commitTimestampMicros,
+          serverTransactionId,
+          isLastRecordInTransaction,
+          recordSequence,
+          tableName,
+          columnTypes,
+          mods,
+          modType,
+          captureType,
+          numberOfRecordsInTransaction,
+          numberOfPartitionsInTransaction,
+          metadata,
+          spannerDatabase,
+          spannerInstanceId);
+    }
+
+    static class Builder {
+      private boolean includeSpannerSource = false;
+      private String spannerDatabase;
+
+      private String spannerInstanceId;
+
+      public Builder setIncludeSpannerSource(Boolean value) {
+        this.includeSpannerSource = value;
+        return this;
+      }
+
+      public Builder setSpannerDatabase(String value) {
+        this.spannerDatabase = value;
+        return this;
+      }
+
+      public Builder setSpannerInstanceId(String value) {
+        this.spannerInstanceId = value;
+        return this;
+      }
+
+      public DataChangeRecordToAvroFn build() {
+        return new DataChangeRecordToAvroFn(this);
+      }
+    }
   }
 
   private static long timestampToMicros(Timestamp ts) {
     return TimeUnit.SECONDS.toMicros(ts.getSeconds())
         + TimeUnit.NANOSECONDS.toMicros(ts.getNanos());
   }
 
-  public static com.google.cloud.teleport.v2.DataChangeRecord dataChangeRecordToAvro(
```

Comment:
It used to be public static, can you comment on why we can not keep public static any more?

Reply:
Because we need to call class functions `spannerDatabaseId()` and `spannerInstanceId()` in this function so it is no longer stateless.